### ISSUE TITLE
Support multiple Allowed Origins in CORS

### ISSUE THREAD

**Issue by straygar** [awslabs/aws-lambda-powertools-python#1006]
**Is your feature request related to a problem? Please describe.**
In some cases, we would like to support multiple origins in CORS when defining an API Gateway/ALB event handler (e.g. support calling an API from "localhost" in dev, or webapps living in different origins calling the same API).

**Describe the solution you'd like**
Allow providing a list of origins `CORSConfig(allow_origins=[])` ([source](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#cors)). If the client's `Origin` is in that list, that value is returned in the response header. Otherwise - an arbitrary (e.g. 1st) value from the list can be returned.

**Comment by michaelbrewer** [comment]
This will rely on using the `origin` to determine which one of the possible `allow_origins` to use. So a little thought will be needed in how to implement this, this article covers this fairly well: https://crashtest-security.com/multiple-values-access-control-allow-origin/

FYI @heitorlessa

**Comment by heitorlessa** [comment]
Hey @straygar thanks a lot for raising this idea with us!

I'd be careful on relying on `Origin` to make that decision for two reasons: **1/** It's an attack vector as this can be easily spoofed, and **2/** API Gateway doesn't support multiple origins by default, which means you need to create a `OPTIONS` method for **each path of your API definition** pointing to your Lambda function handling it.

As 2 can become tedious and error-prone, it opens the door for these sort of attacks that are quite common: https://we45.com/blog/3-ways-to-exploit-cors-misconfiguration

I'd like to wait for more customers demand before we look into it, so we can think of additional factors besides `Origin` and make a good documentation to reduce chances of misconfiguration.

For now, you can manually do it by **a)** configuring API Gateway to invoke your Lambda function for pre-flight requests (`OPTIONS`), **b)** handling pre-flight using `@app.route(rule="*", method="OPTIONS")`, and **c)** dynamically updating `app.cors` with a `CorsConfig` object prior to return (it'll be easier once [middleware feature is finished](#954)

**Comment by michaelbrewer** [comment]
Optimal CORS configuration is always tricky. Classical solutions are a plenty via nginx configurations at allow for a set of regex origins. 

Not sure why there is a gap here for API GW. But it sounds like spinning up multiple endpoints is what is recommended. 

Better to be safe.

**Comment by michaelbrewer** [comment]
@heitorlessa this is what the linked article does. I am just saying we could also do it in behave of the client.

**Comment by michaelbrewer** [comment]
@heitorlessa NOTE using `_cors` may result in breaking changes in the future. Also not sure what the exploit would be of return the main `allow_origin` as the default over `UnauthorizedError`. I am not sure what the abuse would be?

**Comment by michaelbrewer** [comment]
Just for comment i put up a PR:

```python3
from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver, CORSConfig
from aws_lambda_powertools.event_handler.exceptions import UnauthorizedError

# allow_origin - default fall and fall through option
# allow_origins - additional trusted origins
cors_config = CORSConfig(allow_origin="https://stg1.my.domain/", allow_origins=["https://stg2.my.domain/"])
app = ApiGatewayResolver(cors=cors_config)
```

**Comment by leandrodamascena** [comment]
Moving this issue from Ideas to Backlog to start working on it soon.

**Comment by rubenfonseca** [comment]
Note: Function URLs already support multiple origins:

![image](https://user-images.githubusercontent.com/10713/234587486-cc71546a-3bef-4541-be55-afbd6e8684db.png)

**Comment by rubenfonseca** [comment]
Note: API Gateway HTTP (v2) already supports multiple origins:

![image](https://user-images.githubusercontent.com/10713/234588366-0cf739c6-9b67-4775-9fc0-96c3a8479f30.png)

**Comment by rubenfonseca** [comment]
My current understanding of the situation is:

| | API Gateway REST | API Gateway HTTP | ALB | Function URL |
|-|--------|--------|--------|--------|
| CORS supported natively? | ✅ | ✅ | ❌ | ✅ |
| Multiple Origins supported? | ❌ | ✅ | ❌ | ✅ |
| Can use custom Preflight Lambda? | ✅ | ✅ | ✅ | ✅ | 

So it's a little bit of messy situation. We can of course add support to multiple origins when using a custom Lambda Preflight, but it can also be hard to set it up correctly.

If we're going to move forward, I suggest we keep the existing API and add a new `extra_origins: Optional[List[str]]` parameter. 

Then we can add a new issue for Powertools V3 where we combine both arguments into a `allowed_origins: str | List[str]`.

@heitorlessa thoughts?

## Security

Supporting multiple origins should not cause any additional security concerns. Those mainly come from using *dynamic* origins (e.g: regexes). We should also not leak information by returning an origin when one is not passed -- CORS backend should just accept or reject a requested origin.
